proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,1058 @@
+
+* If a module ``M`` splice imports module ``A``, then compiling ``M`` at stage
+  *R* requires compiling module ``A`` at stage *C*.
+
+* If a module ``N`` quote imports module ``B``, then compiling ``N`` at stage
+  *C* requires compiling module ``B`` at stage *R*.
+
+In general, the implementation may choose to support any number of stages. A
+single stage would require that all modules must be compiled such that they can
+be executed during compilation of subsequent modules, as well as at runtime.
+More than two stages are possible to imagine in some cross-compilation
+scenarios. By far the most common case is two stages.  However, the
+specification is expressed in terms of level offsets rather than stages in order
+to keep the language design abstract rather than overfitting to a particular
+arrangement of stages.
+
+
+Module stage offsetting example
+##############################
+
+The interaction between stages and level offsetting can be understood more clearly through an example.
+Module ``A`` splices ``foo`` from module ``B`` which both quotes ``bar`` from module ``C`` and uses ``baz`` from ``D``::
+
+    {-# LANGUAGE ExplicitLevelImports #-}
+    {-# LANGUAGE TemplateHaskell #-}
+    module A where
+    import splice B (foo)
+
+    -- foo can be used within a splice (level -1) because of the splice import (-1).
+    x = $(foo 10)
+
+
+    {-# LANGUAGE ExplicitLevelImports #-}
+    {-# LANGUAGE TemplateHaskell #-}
+    module B where
+    import D (baz)
+    import quote C (bar)
+
+    -- bar can be used within a quote (level +1) because of the quote import (+1)
+    foo x
+      | baz x = [| bar * 2 |]
+      | otherwise = [| bar |]
+
+    module C where
+    bar = 42
+
+    module D where
+    baz 0 = True
+    baz _ = False
+
+
+Now consider compiling ``A`` at stage *R*.
+
+* ``B`` is required at stage *C*, as it is splice imported from ``A`` at *R*.
+* ``C`` is required at stage *R*, as it is quote imported from ``B`` at *C*.
+* ``D`` is required at stage *C*, as it is normally imported from ``B`` at *C*.
+
+Therefore in order to compile ``A`` at *R*, we have performed
+dependency resolution and require ``B`` at *C*, ``C`` at *R* and ``D`` at *C*.
+
+The perhaps curious case is ``D``: is it needed at compile-time or runtime? It
+does not use a splice import, so one could think it is needed at runtime -- but
+here is where the distinction between the import level offset and base stage is
+relevant. ``D`` is only being imported as a dependency of ``B``, which is at *C*
+stage. This makes ``D`` *also* at the *C* stage! Note how ``baz`` is needed
+at compile time just to define ``foo``, which is properly ``splice`` imported.
+
+The levels of all modules in the transitive closure of a ``splice``-imported
+module are offset by -1. Conversely, ``quote`` imports offset the levels by +1,
+thereby making all the levels align correctly.
+
+
+Case Study: ``pandoc``
+######################
```

Comment:
I'd move the case study to the _Effect and Interactions_ section below, as it's evidence that the design meets its goals.